subsys/net/lib/lwm2m/lwm2m_engine.c

```diff
@@ -333,6 +347,22 @@ static void engine_register_obj_inst(struct lwm2m_engine_obj_inst *obj_inst)
 
 static void engine_unregister_obj_inst(struct lwm2m_engine_obj_inst *obj_inst)
 {
+	struct observe_node *obs, *tmp;
+	sys_snode_t *prev_node = NULL;
+
+	/* remove observer instances accordingly */
+	SYS_SLIST_FOR_EACH_CONTAINER_SAFE(
+			&engine_observer_list, obs, tmp, node) {
+		if (obs->path.obj_id != obj_inst->obj->obj_id ||
```

Comment:
This logic would be clearer written as:
(!(obs->path.obj_id == obj_inst->obj->obj_id && obs->path.obj_inst_id == obj_inst->obj_inst_id))

Reply:
A small helper to remove observers from the list would eliminate code duplication here.  It could take params: obj_id and obj_inst_id.  The obj_inst_id could be s32_t and accept -1 as an option meaning to skip checking it (match only obj_id only).